crates/nu-protocol/src/engine/engine_state.rs

```diff
@@ -64,23 +64,29 @@ impl Clone for IsDebugging {
 /// will refer to the corresponding IDs rather than their definitions directly. At runtime, this means
 /// less copying and smaller structures.
 ///
+/// Many of the larger objects in this structure are stored within `Arc` to decrease the cost of
+/// cloning `EngineState`. While `Arc`s are generally immutable, they can be modified using
+/// `Arc::make_mut`, which automatically clones to a new allocation if there are other copies of
+/// the `Arc` already in use, but will let us modify the `Arc` directly if we have the only
+/// reference to it.
+///
 /// Note that the runtime stack is not part of this global state. Runtime stacks are handled differently,
 /// but they also rely on using IDs rather than full definitions.
 #[derive(Clone)]
 pub struct EngineState {
-    files: Vec<(String, usize, usize)>,
-    file_contents: Vec<(Vec<u8>, usize, usize)>,
+    files: Vec<(Arc<String>, usize, usize)>,
+    file_contents: Vec<(Arc<Vec<u8>>, usize, usize)>,
```

Comment:
Putting those beauties of a type on my future TODO list 😄

Reply:
Haha, yup. Very self-describing

Reply:
There's potentially a bit of a benefit that could be had here by switching from `Arc<String>` to `Arc<str>`, and `Arc<Vec<u8>>` to `Arc<[u8]>` fyi, I just thought it would be more pain than it's worth. But it would remove one indirect pointer reference and we really never need to modify these after the fact